src/common/dao/immutable.go

```diff
@@ -0,0 +1,84 @@
+package dao
+
+import (
+	"fmt"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common"
+	"github.com/goharbor/harbor/src/common/models"
+)
+
+// CreateImmutableRule creates the Immutable Rule
+func CreateImmutableRule(ir *models.ImmutableRule) (int64, error) {
+	ir.Enabled = true
+	if exceedMaxImmutableRuleCount(ir.ProjectID) {
+		return 0, fmt.Errorf("immutable tag rule count exceed %v", common.MaxRuleCount)
+	}
+	o := GetOrmer()
+	return o.Insert(ir)
+}
+
+func exceedMaxImmutableRuleCount(projectID int64) bool {
+	o := GetOrmer()
+	var count int
+	o.Raw("select  count(*) cnt from immutable_tag_rule where project_id = ? ", projectID).QueryRow(&count)
+	return count >= common.MaxRuleCount
+}
+
+// UpdateImmutableRule update the immutable rules
+func UpdateImmutableRule(projectID int, ir *models.ImmutableRule) (int64, error) {
+	o := GetOrmer()
+	return o.Update(ir, "RepoFilter", "TagFilter")
+}
+
+// ToggleImmutableRule enable/disable immutable rules
+func ToggleImmutableRule(id int64, enabled bool) (int64, error) {
+	o := GetOrmer()
+	ir := &models.ImmutableRule{ID: id, Enabled: enabled}
+	return o.Update(ir, "Enabled")
+}
+
+// GetImmutableRule get immutable rule
+func GetImmutableRule(id int64) (*models.ImmutableRule, error) {
+	o := GetOrmer()
+	ir := &models.ImmutableRule{ID: id}
+	err := o.Read(ir)
+	if err == orm.ErrNoRows {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	return ir, nil
+}
+
+// QueryImmutableRuleByProjectID get all immutable rule by project
+func QueryImmutableRuleByProjectID(projectID int64) ([]models.ImmutableRule, error) {
+	o := GetOrmer()
+	qs := o.QueryTable(&models.ImmutableRule{}).Filter("ProjectID", projectID)
+	var r []models.ImmutableRule
+	_, err := qs.All(&r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get immutable tag rule by projectID %d, error: %v", projectID, err)
+	}
+	return r, nil
+}
+
+// QueryEnabledImmutableRuleByProjectID get all enabled immutable rule by project
+func QueryEnabledImmutableRuleByProjectID(projectID int64) ([]models.ImmutableRule, error) {
```

Comment:
Merging `QueryImmutableRuleByProjectID ` and `QueryEnabledImmutableRuleByProjectID ` as `QueryImmutableRules(q *Query)`, the query conditions can be specified in the `Query` object

Reply:
One API is query all, another API is just query enabled rule